tensorflow/compiler/tf2tensorrt/convert/utils.h

```diff
@@ -128,77 +129,199 @@ inline bool IsTrtShapeTensorCompatible(const Tensor& tensor) {
-                             std::vector<PartialTensorShape>* input_shapes);
+  // Convert to an nvinfer1::Dims and return by value.
+  StatusOr<nvinfer1::Dims> AsTrtDims() const {
+    nvinfer1::Dims result;
+    TF_RETURN_IF_ERROR(TrtDims(&result));
+    return result;
+  }
 
-// Creates PartialTensorShape from nvinfer1::Dims. The "batch_dimension", when
-// provided, is prepended to the shape.
-inline PartialTensorShape TrtDimsToPartialTensorShape(
-    const nvinfer1::Dims& dims,
-    absl::optional<int64> batch_dimension = absl::nullopt) {
-  absl::InlinedVector<int64, 4> dims_vec;
-  if (batch_dimension) {
-    dims_vec.insert(dims_vec.begin(), std::max(int64(-1), *batch_dimension));
-  }
-  std::copy(dims.d, dims.d + dims.nbDims, std::back_inserter(dims_vec));
-  return PartialTensorShape(dims_vec);
-}
+  // Convert to a TensorShape.
+  Status TensorShape(TensorShape* shape,
+                     absl::optional<int> batch_size = absl::nullopt) const {
+    TF_RETURN_IF_ERROR(status_);
+    TF_RETURN_IF_ERROR(TensorShapeUtils::MakeShape(
+        static_cast<const int64_t*>(storage_.data()), storage_.size(), shape));
+    if (batch_size) shape->InsertDim(0, *batch_size);
+    return Status::OK();
+  }
 
-Status TrtDimsToTensorShape(const std::vector<int>& trt_dims,
-                            TensorShape* shape,
-                            absl::optional<int> batch_size = absl::nullopt);
+  Status PartialTensorShape(
+      PartialTensorShape* shape,
+      absl::optional<int> batch_size = absl::nullopt) const {
+    TF_RETURN_IF_ERROR(status_);
+    TF_RETURN_IF_ERROR(TensorShapeUtils::MakeShape(
+        static_cast<const int64_t*>(storage_.data()), storage_.size(), shape));
+    if (batch_size) shape->InsertDim(0, *batch_size);
+    return Status::OK();
+  }
 
-template <typename TensorShapeType>
-Status TrtDimsToTensorShape(const nvinfer1::Dims trt_dims,
-                            TensorShapeType* shape,
-                            absl::optional<int> batch_size = absl::nullopt) {
-  TF_RETURN_IF_ERROR(
-      TensorShapeUtils::MakeShape(trt_dims.d, trt_dims.nbDims, shape));
-  if (batch_size) {
-    shape->InsertDim(0, batch_size.value());
-  }
-  return Status::OK();
-}
+  Status Vector(std::vector<int32_t>* shape) const {
+    TF_RETURN_IF_ERROR(status_);
+    shape->clear();
+    absl::c_copy(storage_, std::back_inserter(*shape));
+    return Status::OK();
+  }
+
+  //----- Property Accessors ------
+
+  // Returns true if the shape should be interpreted as a scalar. This follows
+  // TensorRT conversions: a scalar shape can have NumDims()==1 or NumDims()==0,
+  // but the underlying storage_ container has a single dimension of size 1.
+  bool IsStatic() const {
+    return !absl::c_any_of(storage_, [](auto i) { return i < 0; });
+  }
+
+  // Returns product of all dimensions.
+  int64_t Volume() const {
+    return absl::c_accumulate(storage_, int64_t(1), std::multiplies<>());
```

Comment:
Should the status_ field play a role here?

Reply:
It can... I was trying to strike a middle ground between introducing too many changes and introducing additional checks that the current code doesn't have. If you OK with it basically adding 1-2 additional lines at call sites for the check, then I will update it

Reply:
Will update to `StatusOr<bool>`